File: backend/src/main/java/com/taskflow/kanban/workspace/service/WorkspaceService.java
package com.taskflow.kanban.workspace.service;

import com.taskflow.kanban.workspace.dto.WorkspaceCreateDto;
import com.taskflow.kanban.workspace.dto.WorkspaceDto;
import com.taskflow.kanban.workspace.dto.WorkspaceMemberDto;
import com.taskflow.kanban.workspace.dto.WorkspaceUpdateDto;
import com.taskflow.kanban.workspace.entity.WorkspaceRole;

import java.util.List;
import java.util.UUID;

public interface WorkspaceService {
    WorkspaceDto createWorkspace(WorkspaceCreateDto createDto);
    WorkspaceDto getWorkspaceById(UUID id);
    List<WorkspaceDto> getAllWorkspacesForCurrentUser();
    WorkspaceDto updateWorkspace(UUID id, WorkspaceUpdateDto updateDto);
    void deleteWorkspace(UUID id);
    List<WorkspaceMemberDto> getMembers(UUID workspaceId);
    void addMember(UUID workspaceId, UUID userId, WorkspaceRole role);
    void removeMember(UUID workspaceId, UUID userId);
    void updateMemberRole(UUID workspaceId, UUID userId, WorkspaceRole role);
}

File: frontend/src/app/core/services/workspace.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Workspace, WorkspaceCreateDto, WorkspaceUpdateDto } from '../models/workspace.model';
import { WorkspaceMember, WorkspaceRole, WorkspaceMemberUpdateDto } from '../models/workspace-member.model';

@Injectable({
  providedIn: 'root'
})
export class WorkspaceService {
  private readonly API_URL = `${environment.apiUrl}/workspaces`;

  constructor(private http: HttpClient) {}

  getAllWorkspaces(): Observable<Workspace[]> {
    return this.http.get<Workspace[]>(this.API_URL);
  }

  getWorkspaceById(id: string): Observable<Workspace> {
    return this.http.get<Workspace>(`${this.API_URL}/${id}`);
  }

  createWorkspace(workspace: WorkspaceCreateDto): Observable<Workspace> {
    return this.http.post<Workspace>(this.API_URL, workspace);
  }

  updateWorkspace(id: string, workspace: WorkspaceUpdateDto): Observable<Workspace> {
    return this.http.put<Workspace>(`${this.API_URL}/${id}`, workspace);
  }

  deleteWorkspace(id: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  getMembers(workspaceId: string): Observable<WorkspaceMember[]> {
    return this.http.get<WorkspaceMember[]>(`${this.API_URL}/${workspaceId}/members`);
  }

  addMember(workspaceId: string, userId: string): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/${workspaceId}/members/${userId}`, {});
  }

  removeMember(workspaceId: string, userId: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${workspaceId}/members/${userId}`);
  }

  updateMemberRole(workspaceId: string, userId: string, role: WorkspaceRole): Observable<void> {
    const updateDto: WorkspaceMemberUpdateDto = { role };
    return this.http.put<void>(`${this.API_URL}/${workspaceId}/members/${userId}`, updateDto);
  }
}